Source/iOS/OIDExternalUserAgentIOS.m

```diff
@@ -65,65 +65,75 @@ - (BOOL)presentExternalUserAgentRequest:(id<OIDExternalUserAgentRequest>)request
-                                 description:nil];
-        [strongSelf->_session failExternalUserAgentFlowWithError:safariError];
-      }
-    }];
-    _webAuthenticationVC = authenticationVC;
-    openedSafari = [authenticationVC start];
+    if (!UIAccessibilityIsGuidedAccessEnabled()) {
+      __weak OIDExternalUserAgentIOS *weakSelf = self;
+      NSString *redirectScheme = request.redirectScheme;
+      ASWebAuthenticationSession *authenticationVC =
+          [[ASWebAuthenticationSession alloc] initWithURL:requestURL
+                                        callbackURLScheme:redirectScheme
+                                         completionHandler:^(NSURL * _Nullable callbackURL,
+                                                             NSError * _Nullable error) {
+        __strong OIDExternalUserAgentIOS *strongSelf = weakSelf;
+        if (!strongSelf) {
+            return;
+        }
+        strongSelf->_webAuthenticationVC = nil;
+        if (callbackURL) {
+          [strongSelf->_session resumeExternalUserAgentFlowWithURL:callbackURL];
+        } else {
+          NSError *safariError =
+              [OIDErrorUtilities errorWithCode:OIDErrorCodeUserCanceledAuthorizationFlow
+                               underlyingError:error
+                                   description:nil];
+          [strongSelf->_session failExternalUserAgentFlowWithError:safariError];
+        }
+      }];
+      _webAuthenticationVC = authenticationVC;
+      openedSafari = [authenticationVC start];
+    }
+  }
   // iOS 11, use SFAuthenticationSession
-  } else if (@available(iOS 11.0, *)) {
-    __weak OIDExternalUserAgentIOS *weakSelf = self;
-    NSString *redirectScheme = request.redirectScheme;
-    SFAuthenticationSession *authenticationVC =
-        [[SFAuthenticationSession alloc] initWithURL:requestURL
-                                   callbackURLScheme:redirectScheme
-                                   completionHandler:^(NSURL * _Nullable callbackURL,
-                                                       NSError * _Nullable error) {
-      __strong OIDExternalUserAgentIOS *strongSelf = weakSelf;
-      if (!strongSelf) {
-          return;
-      }
-      strongSelf->_authenticationVC = nil;
-      if (callbackURL) {
-        [strongSelf->_session resumeExternalUserAgentFlowWithURL:callbackURL];
-      } else {
-        NSError *safariError =
-            [OIDErrorUtilities errorWithCode:OIDErrorCodeUserCanceledAuthorizationFlow
-                             underlyingError:error
-                                 description:@"User cancelled."];
-        [strongSelf->_session failExternalUserAgentFlowWithError:safariError];
-      }
-    }];
-    _authenticationVC = authenticationVC;
-    openedSafari = [authenticationVC start];
+  // SFAuthenticationSession doesn't work with guided access (rdar://40809553)
+  if (@available(iOS 11.0, *)) {
+    if (!openedSafari && !UIAccessibilityIsGuidedAccessEnabled()) {
```

Comment:
I don't see how `openedSafari` can ever be true at this point is that check needed?

Reply:
True, I will fix that.

Reply:
Actually it is needed in this context indeed: if on iOS12 not in access mode, `openedSafari` will have the result of `[authenticationVC start]`. If it did not work with ASWebAuthenticationSession for some reason, it will still try SFAuthenticationSession.
What do you think? Is there a better way to handle this?

Reply:
Ah ok, I missed that the `else if` was removed.

The new logic is a little harder to read, but it's also inconsistent. In most places "openedSafari" is the test rather than the platform version, but in the case of mobile Safari there's no `openedSafari` check, and happens only for iOS 8 & below like before (mobile safari is opened in the `else` path of the iOS 9 version check).

I think we need to holistically look at the logic here and move to consistent model, not just apply a fix for UIAccessibilityIsGuidedAccessEnabled.

Perhaps that consistent model is "did the previous preferred method work, otherwise try the next best option", which is almost what you've done here, with the exception of mobile Safari.

Reply:
Makes sense. I didn't want to modify the code too much for that fix, but you are right. I will make the changes.

Reply:
I added the condition on the `else` path. Do you think we need a larger refactor to this model?

Reply:
I think it's good to go. The approach seems valid, it's basically a cascading order of preferences that's less attached to the iOS version now (while still gated by the version).

Reply:
Can you address the nit regarding the comment, and squish the commits into 1?

Reply:
You don't squash for merging pull requests?